docs: Add blog post on how we test vector (#1363)
timberio/vector: website/blog/2020-02-12-how-we-test-vector.md

A unit test is generally defined by the idea of isolation. This can mean
different things to different people, but a common definition is that the test
is both isolated from the outside world (i.e. no network calls, reading files,
etc) and exercises a single component of the system (e.g. one function, class,
etc).

In Vector, we've found unit tests to be a particularly great fit for our
transforms. This makes sense, since transforms are effectively isolated
functions that take events in and return events out. For the same reason, we
tend to use unit tests extensively around the encoder portions of our sinks.

For other types of components, so much of the functionality is focused on
communication with external systems that it can be difficult to isolate logic
enough for a unit test. As much as possible, we try to critically assess that
difficulty and use it to help find ways we can refactor the design to be more
modular and therefore amenable to unit testing. Unit tests are an excellent
source of design feedback, so we want to feel that pain and refactor when
they're difficult to write.

That being said, there are two specific places we often run into the limitations

> of unit tests. The first and more obvious is around pieces of code that are

Reviewer: I often prefer to embolden things. :) Eg `**fundamentally not isolated**` and `**size of the input space** and **number of potential paths**`. But both work great. Having am emphasis on them will help them stand out.

Takeaways:

* Isolation makes unit tests simple, fast, and reliable.
* If something is difficult to unit test, refactor until it's easy.
* As a human, be wary of your ability to think up exhaustive example inputs.

### Integration tests

The category of integration tests is a bit of a catch-all. Roughly defined,
they're example-based tests that are explicitly not isolated and focus on the
interaction between one or more components.

Given Vector's focus on integrating with a wide variety of external systems, we
have a higher ratio integration tests than your average system. Even once we've
done all we can to isolate logic into small, unit-testable functions, we still
need to ensure the component as a whole does what it's supposed to. If unit
tests tell you that something works in theory, integration tests tell you that
it _should_ work in practice.

Similar to unit tests, there are two major downsides to integration tests. The
first is the same problem unit tests have with exhaustiveness, but intensely
magnified. Because they often cover the interaction of multiple whole systems,
integration tests will virtually never cover the combinatorial explosion of

> potential execution paths. Second, integration tests are often simply a pain to

Reviewer: ```suggestion
integration tests will virtually never cover all possible execution paths. Second, integration tests are often simply a pain to
```

write and maintain. Given their dependence on external systems, they can be
tedious to write, slow to run, and often flaky if their environment isn't set up
just so.

Takeaways:

* While they can be a pain, integration tests are a valuable sanity check that
  your system will actually work for users.

> * Don't overuse them trying to cover every possible scenario, because you can't.

Reviewer: I suggest adding

```markdown
* Integration tests may rely on external services or components.
```

## Generative testing

The idea of generative testing is a direct response to the shared shortcoming of
example-based strategies like unit and integration testing. Humans are bad at
visualizing very large state spaces, and the state space of your code's possible
inputs and execution paths is enormous. You also bring all the same biases when
writing tests as you did when writing the original code.

Generative tests address these issues by taking the human out of the equation
and having the computer generate millions of example inputs. The tradeoff is
that since you aren't simply hardcoding a list of inputs and their expected
outputs anymore, you're forced to come up with more creative ways of identifying

> failures.

Reviewer: ```suggestion
outputs anymore, you're often forced to come up with creative ways of identifying
failures.
```

things like throughput and system resource use, we instead run each
configuration through different interesting scenarios to see how they behave.

For example, we have correctness tests around various flavors of file rotation,
disk persistence across restarts, nested JSON messages, etc. While these are
behaviors that we also test at various lower levels (e.g. unit and integration
tests), covering a handful of important cases at this level of abstraction gives
us some extra confidence that we are seeing exactly what our users will see.

The ability to compare behaviors across competing tools is another bonus. Going
through the process of setting up those tests gets us valuable experience
working with those other tools. We can see what works well in their
configuration, documentation, etc, and identify areas where we can improve
Vector as a result.

Takeaways:

* Taking time to "zoom out" and test your system as a user would can help
  uncover blind spots and sanity-check behavior.
* Evaluating similar tools can help build a better understanding of user
  expectations.

### Reliability tests

A third category that we're currently working to integrate into

> `vector-test-harness` is something we're calling reliability tests. These are

Reviewer: Link to this!

```suggestion
[`vector-test-harness`](https://github.com/timberio/vector-test-harness) is something we're calling reliability tests. These are
```

reliability test of our S3 sink exposed a bug where a specific kind of network
failure could lead to duplicate data when the retried request crossed
a timestamp boundary. That is not the type of failure we expect to be able to
induce in a local integration test, and the relevant factors (time and network
conditions) were not those exercised by standard fuzzing or property-based
testing.

The main challenge with these kinds of tests, aside from getting the requisite
environment and harnesses up and running, is capturing sufficient context about
the environment at the time of the failure that you stand a chance at
understanding and reproducing it. This task itself is a great test for our
internal observability, and any issue we can't reproduce is a sign that our
logging and metrics data needs to be improved.

Another issue with these tests is that the vast majority of the time, nothing
particularly interesting is happening. Since we want to find bugs as quickly as
possible, we can supplement the randomness of the environment by injecting
various types of faults on our own. There are a variety of tools for this, such

> as [Toxiproxy][10] and [Namazu][11].

Reviewer: Might want to mention `tc` since it's on most Linux hosts already. :)
